x-pack/test/security_solution_endpoint/apps/endpoint/endpoint_transform.ts

```diff
@@ -0,0 +1,147 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { Client } from '@elastic/elasticsearch';
+import expect from '@kbn/expect';
+import { FtrProviderContext } from '../../configs/ftr_provider_context';
+import { targetTags } from '../../target_tags';
+
+export default ({ getPageObjects, getService }: FtrProviderContext) => {
+  const retry = getService('retry');
+  const endpointTestResources = getService('endpointTestResources');
+  const esClient: Client = getService('es');
+  const sleep = (ms: number) => new Promise((f) => setTimeout(f, ms));
+
+  const transformAggregation = {
+    init_script: 'state.docs = []',
+    map_script: "state.docs.add(new HashMap(params['_source']))",
+    combine_script: 'return state.docs',
+    reduce_script:
+      "def ret = new HashMap(); for (s in states) { for (d in s) { if (d.containsKey('Endpoint')) { ret.endpoint = d } else { ret.agent = d } }} return ret",
+  };
+
+  describe('endpoint transforms', function () {
+    targetTags(this, ['@ess', '@serverless', '@serverlessQA']);
```

Comment:
Are we sure we want this test to be a part of release pipeline(`@serverlessQA`)? I would suggest to observe it some in the periodic pipeline and promote it only when we are sure it is safe to do so ;)

Reply:
This test was specifically requested to be created for the release pipeline, since our existing tests that cover this functionality were a little too flaky

But I do hear your point of observing first. The test is quite static, in that its making two indices from scratch, running a transform, looking at the output. Nearly the whole world around it could change and I think the test would still pass. does not rely on the existence of the fleet or even security solution plugin. Just testing a raw elasticsearch feature (which was the goal, there was a change to transforms there)

Reply:
I've manually triggered a periodic pipeline. Let's see if everything works as expected. - https://buildkite.com/elastic/kibana-serverless-security-solution-quality-gate-defend-workflows/builds/853#0190b6ac-51f8-4ef3-a076-15a6c9a46398

Reply:
@pzl ,

> "...does not rely on the existence of the fleet or even security solution plugin. Just testing a raw elasticsearch feature (which was the goal, there was a change to transforms there)"

if thats the case, why is the test under security solution?  Should  this be a test of Core or ES?


____

I'm assuming this is attempting to test that one of our transforms works as expected? if so, could we actually use the respective package that installs it  and then test with a the real thing?

Reply:
> why is the test under security solution

Sure, there is a little bit of an inverted premise here. We are essentially testing and catching any changes _in elasticsearch_. Presumably, the tests would be there.

But that would mean in ES, it's testing for every solution that uses this (I think there are 4-5 exceptions that were added). So these would be tests in ES about 5 different kibana plugins to make sure it isn't breaking these downstream users. 

It seemed easier to have the solutions make sure they have (stable) tests that verify their transform usage works as intended, and add them as quality gates. We 1 of 5 doing that.

We are verifying that _security solution_ is able to use our regular transform and nothing upstream is blocking that functionality.

> could we actually use the respective package

package install, sometimes flaky, relies on fleet things, tests a whole lot of functionality in that whole lifecycle we don't need. This is a quality gate just to verify upstream (es) didn't accidentally break our transform. So this was intentionally kept to minimal dependency. Designed to only test one thing, the operation of a scripted metric aggregation transform

Reply:
++, it's important that there's a test that proves that ES doesn't break a solution down stream; having a test in ES would have to be tautological ("given that we want this query to run, we'll produce the right output"), since it can't test the actual Security Solution's code ("given ES from main, performing the transform still works").